### ISSUE TITLE
Issue removing DRM from Kindle book

### ISSUE THREAD

**Issue by tmthywynn8** [noDRM/DeDRM_tools#380]
### Question / bug report

Looks like there are still some Python 2 syntax errors in this version of DeDRM, as a modified version of v6.8.0 works fine (see below):

```log
DeDRM v6.8.0: Trying to decrypt {ASIN}_EBOK.azw
Using Library AlfCrypto DLL/DYLIB/SO
MobiDeDrm v0.42.
Copyright © 2008-2017 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: {title}
Got DSN key from database default_key
Got DSN key from database default_key_2
Found 8 keys to try after 0.1 seconds
Crypto Type is: 2
File is encoded with PID {PID}
Decrypting. Please wait . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . done
Decryption succeeded after 0.3 seconds
HDImage Container file is found: {name}.azw.res
HDImage merge start...
HDImage merge succeeded after 2.6 seconds
DeDRM v6.8.0: Finished after 2.6 seconds
```

### Which version of Calibre are you running?

2.85.1

### Which version of the DeDRM plugin are you running?

v10.0.9

### If applicable, which version of the Kindle software are you running?

1.26.0

### Log output

```log
DeDRM v10.0.9: Trying to decrypt {ASIN}_EBOK.azw
MobiDeDrm v1.1.
Copyright © 2008-2022 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: {title}
Got DSN key from database default_key_2
Error getting PIDs from database default_key_2: unsupported operand type(s) for >>: 'str' and 'int'
Traceback (most recent call last):
  File "calibre_plugins.dedrm.kgenpids", line 299, in getPidList
  File "calibre_plugins.dedrm.kgenpids", line 271, in getK4Pids
  File "calibre_plugins.dedrm.kgenpids", line 100, in encodePID
  File "calibre_plugins.dedrm.kgenpids", line 92, in getSixBitsFromBitField
  File "calibre_plugins.dedrm.kgenpids", line 87, in getTwoBitsFromBitField
TypeError: unsupported operand type(s) for >>: 'str' and 'int'
Got DSN key from database default_key
Error getting PIDs from database default_key: unsupported operand type(s) for >>: 'str' and 'int'
Traceback (most recent call last):
  File "calibre_plugins.dedrm.kgenpids", line 299, in getPidList
  File "calibre_plugins.dedrm.kgenpids", line 271, in getK4Pids
  File "calibre_plugins.dedrm.kgenpids", line 100, in encodePID
  File "calibre_plugins.dedrm.kgenpids", line 92, in getSixBitsFromBitField
  File "calibre_plugins.dedrm.kgenpids", line 87, in getTwoBitsFromBitField
TypeError: unsupported operand type(s) for >>: 'str' and 'int'
Found 0 keys to try after 0.3 seconds
Crypto Type is: 2
DeDRM v10.0.9: Failed to decrypt with error: unsupported operand type(s) for +: 'int' and 'str'
DeDRM v10.0.9: Looking for new default Kindle Key after 0.3 seconds
searching for kinfoFiles in %localappdata%
Found K4PC 1.25+ kinf2018 file: %localappdata%\Amazon\Kindle\storage\.kinf2018
DeDRM v10.0.9: Exception when getting default Kindle Key after 0.4 seconds
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 943, in KindleMobiDecrypt
  File "calibre_plugins.dedrm.kindlekey", line 894, in kindlekeys
  File "calibre_plugins.dedrm.kindlekey", line 382, in getDBfromFile
  File "calibre_plugins.dedrm.kindlekey", line 382, in <dictcomp>
  File "calibre_plugins.dedrm.kindlekey", line 127, in encodeHash
  File "calibre_plugins.dedrm.kindlekey", line 119, in encode
TypeError: unsupported operand type(s) for ^: 'str' and 'int'
DeDRM v10.0.9: Ultimately failed to decrypt after 0.4 seconds. Read the FAQs at noDRM's repository: https://github.com/noDRM/DeDRM_tools/blob/master/FAQs.md
```

**Comment by ElleKayEm** [comment]
Is this on Mountain Lion?  Is that why you're using calibre 2.85.1?

**Comment by noDRM** [comment]
Can you give this version a try and see if that works, and if not, post a new log?

https://github.com/noDRM/DeDRM_tools/suites/14785762000/artifacts/841182896

**Comment by tmthywynn8** [comment]
I'm leaving the PIDs in because it has odd output. Hopefully that's not sensitive information.

```log
DeDRM v10.0.9: Trying to decrypt {ASIN}_EBOK.azw
MobiDeDrm v1.1.
Copyright © 2008-2022 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: {title}
Got DSN key from database default_key
Got DSN key from database default_key_2
Found 8 keys to try after 0.4 seconds
Crypto Type is: 2
Warning: PID 9oe8Ocjv['C']['D'] has wrong number of digits
Warning: PID ['E']['T']['T']['X']['U']['5']['E']['C']['V']['5'] has wrong number of digits
Warning: PID vJD+GDv3['3']['B'] has wrong number of digits
Warning: PID ['4']['K']['X']['H']['N']['N']['B']['G']['D']['4'] has wrong number of digits
Warning: PID l9+3qHfC['I']['7'] has wrong number of digits
Warning: PID ofw2i+ns['1']['G'] has wrong number of digits
Warning: PID 058lHoPs['N']['R'] has wrong number of digits
Warning: PID PM/TIuCO['V']['G'] has wrong number of digits
DeDRM v10.0.9: Failed to decrypt with error: unsupported operand type(s) for +: 'int' and 'str'
DeDRM v10.0.9: Looking for new default Kindle Key after 0.4 seconds
searching for kinfoFiles in %localappdata%
Found K4PC 1.25+ kinf2018 file: %localappdata%\Amazon\Kindle\storage\.kinf2018
DeDRM v10.0.9: Exception when getting default Kindle Key after 0.4 seconds
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 943, in KindleMobiDecrypt
  File "calibre_plugins.dedrm.kindlekey", line 900, in kindlekeys
  File "calibre_plugins.dedrm.kindlekey", line 415, in getDBfromFile
  File "calibre_plugins.dedrm.kindlekey", line 243, in GetUserName
TypeError: one character unicode string expected
DeDRM v10.0.9: Ultimately failed to decrypt after 0.4 seconds. Read the FAQs at noDRM's repository: https://github.com/noDRM/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "site-packages\calibre\customize\ui.py", line 170, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 1030, in run
  File "calibre_plugins.dedrm.__init__", line 981, in KindleMobiDecrypt
DeDRMError: DeDRM v10.0.9: Ultimately failed to decrypt after 0.4 seconds. Read the FAQs at noDRM's repository: https://github.com/noDRM/DeDRM_tools/blob/master/FAQs.md
```

**Comment by noDRM** [comment]
Darn, looks like I missed some places.
Can you give this version a try and post a new log?

https://github.com/noDRM/DeDRM_tools/suites/14801809295/artifacts/842359792

**Comment by tmthywynn8** [comment]
It took quite a bit longer than 6.8.0, but seems to have worked perfectly. Thanks for taking the time to support EOL versions of Calibre. It really is appreciated.

```log
DeDRM v10.0.9: Trying to decrypt {ASIN}_EBOK.azw
MobiDeDrm v1.1.
Copyright © 2008-2022 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: {title}
Got DSN key from database default_key
Got DSN key from database default_key_2
Found 8 keys to try after 0.3 seconds
Crypto Type is: 2
File is encoded with PID {PID}.
Decrypting. Please wait . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . done
Decryption succeeded after 12.6 seconds
DeDRM v10.0.9: Finished after 12.6 seconds
```

**Comment by noDRM** [comment]
Thanks for testing and confirming it now works.

The reason it takes a bit longer is that all the decryption happens in Python code. The old versions had a bunch of code to hand off heavy encryption tasks to native OpenSSL code running outside of Python (which is faster), while in the newer versions I dropped that in favour of Python code - which is slower but way easier to maintain and less error-prone. I figured nobody is going to import hundreds of DRMed books at the same time so it doesn't matter if it's a bit slower.
